ShiftViT Example Enhancements

shivalikasingh95 opened the pull request:
Hi, I have updated the ShiftViT example with links to model & space on Hugging Face Hub.

While working on setting up a demo space for ShiftViT, I came across a few errors while saving and loading the model.
I have added fixes for those errors as well as part of this PR.

I have also added a section on Model Inference for the benefit of the users of Keras as I felt that was missing from the example.
However, if you feel the section is not needed then do let me know, I'll revert those changes.

fchollet commented:
Thanks for the PR!

> Continuous integration / black (pull_request) Failing after 17s — black

Please fix.

fchollet reviewed (commented):
Please limit the PR to just the Python file for now.

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -681,6 +727,15 @@ def test_step(self, data):
         self.compiled_metrics.update_state(labels, logits)
         return {m.name: m.result() for m in self.metrics}
 
+    def call(self, images):
```
While saving the model, I was getting this error:
_"Model <__main__.ShiftViTModel object at 0x7f697714ce10> cannot be saved either because the input shape is not available or because the forward pass of the model is not defined.To define a forward pass, please override `Model.call()`. To specify an input shape, either call `build(input_shape)` directly, or call the model on actual data using `Model()`, `Model.fit()`, or `Model.predict()`. If you have a custom training step, please make sure to invoke the forward pass in train step through `Model.__call__`, i.e. `model(inputs)`, as opposed to `model.call()"_

Hence, defined the forward pass of the model

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -784,11 +839,25 @@ def __call__(self, step):
+# pass sample data to the model so that input shape is available at the time of
+# saving the model using tf.keras.models.save_model()
+sample_ds, _ = next(iter(train_ds))
+model(sample_ds, training=False).shape
```
Again, was getting this error while saving model:
"Model <main.ShiftViTModel object at 0x7f697714ce10> cannot be saved either because the input shape is not available"

Hence, added this fix

fchollet replied:
Remove the `.shape` above, it's not needed

shivalikasingh95 replied:
sure, it's removed now

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -840,6 +909,125 @@ def __call__(self, step):
+tf.keras.models.save_model(model, "./ShiftViT")
+
+"""
+## Model inference
```
Added a section on Model inference as I felt it would make the example more end-to-end and usable.

Please let me know if this section needs any fixes or if it can be made better in any way.

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -784,11 +839,25 @@ def __call__(self, step):
             step > self.total_steps, 0.0, learning_rate, name="learning_rate"
         )
 
+    def get_config(self):
```
Added this to avoid following error while saving model:
_"NotImplementedError: Learning rate schedule must override get_config"_

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -543,6 +566,24 @@ def call(self, x, training=False):
 
+    # Since this is a custom layer, we need to overwrite get_config()
+    # so that model can be easily saved & loaded after training
+    def get_config(self):
```
Fix for this error while saving model:

_"NotImplementedError: 
Layer StackedShiftBlocks has arguments ['self', 'epsilon', 'mlp_dropout_rate', 'num_shift_blocks', 'stochastic_depth_rate', 'is_merge', 'num_div', 'shift_pixel', 'mlp_expand_ratio']
in `__init__` and therefore must override `get_config()`"_

shivalikasingh95 commented:
Hi @fchollet ,
I've fixed this: "Continuous integration / black (pull_request) Failing after 17s — black"
I've also removed the generated files and limited the PR to only the Python file now.

Have also added some comments explaining why I made certain changes to the code. Looking forward to your review.

fchollet commented on examples/vision/shiftvit.py:
```diff
@@ -840,6 +909,125 @@ def __call__(self, step):
+
+It can be saved in TF SavedModel format only. In general, this is the recommended format for saving models as well.
+"""
+tf.keras.models.save_model(model, "./ShiftViT")
```
Just do `model.save(...)`

shivalikasingh95 replied:
this is done too

fchollet reviewed (commented):
Overall things look good! How many lines of code are that in total? (try to run the `add_example` command to see). You might want to make to a bit simpler / shorter to fit within the 300 line limit.

shivalikasingh95 commented:
> Overall things look good! How many lines of code are that in total? (try to run the `add_example` command to see). You might want to make to a bit simpler / shorter to fit within the 300 line limit.

Hi @fchollet the original example itself was > 300 lines (403 lines to be precise). This was called out by the original author while adding the example too as mentioned [here](https://github.com/keras-team/keras-io/pull/811#issuecomment-1054963477) 

After my current changes, the lines of code has increased to 497.
Let me know if this would be okay or a blocker.

fchollet reviewed (approved):
Not a blocker, that's ok, we can make an exception. Please add the generated files.

shivalikasingh95 commented on examples/vision/shiftvit.py:
```diff
@@ -1,8 +1,8 @@
 """
 Title: A Vision Transformer without Attention
-Author: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498), [Ritwik Raha](https://twitter.com/ritwik_raha)
+Author: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498), [Ritwik Raha](https://twitter.com/ritwik_raha), [Shivalika Singh](https://www.linkedin.com/in/shivalika-singh/)
```
@fchollet have updated the author list after checking with @ariG23498 & @ritwikraha 
Hope this is fine.

ariG23498 replied:
Absolutely fine by me. Keep up the good work!

shivalikasingh95 replied:
Thank you so much :) @ariG23498

shivalikasingh95 commented:
> Not a blocker, that's ok, we can make an exception. Please add the generated files.

Thanks a lot @fchollet ! 
Also apologies for the delay at my end. Have updated the PR with generated files now. In case, anything needs to be updated please let me know.

shivalikasingh95 commented:
@fchollet would appreciate if this PR could be merged if it looks good to go.

I have [another PR](https://github.com/keras-team/keras-io/pull/1015) with minimal changes that needs a quick review. If that looks good, I can add the generated files for it.

Would be really grateful if you do a quick check for both the PRs whenever you find the time.

shivalikasingh95 commented:
Sorry to trouble you again @fchollet. I'm not sure who are the other maintainers for this repository.

But it would be great if you/another maintainer could take a quick look at this PR as well as my [other PR](https://github.com/keras-team/keras-io/pull/1015). Would be grateful if both can be closed soon if they look good to go.

pcoet commented:
Hi @shivalikasingh95, thanks for the PR. It looks like @fchollet approved this, and you generated the markdown and notebook files. I'll go ahead and merge. Thanks!
